NetKet/Machine/fullconn_layer.hpp

```diff
@@ -0,0 +1,256 @@
+                        << out_size_ << std::endl;
+    InfoMessage(buffer) << "# # UseBias = " << usebias_ << std::endl;
+  }
+
+  void to_json(json &pars) const override {
+    json layerpar;
+    layerpar["Name"] = "FullyConnected";
+    layerpar["UseBias"] = usebias_;
+    layerpar["Inputs"] = in_size_;
+    layerpar["Outputs"] = out_size_;
+    layerpar["Bias"] = bias_;
+    layerpar["Weight"] = weight_;
+
+    pars["Machine"]["Layers"].push_back(layerpar);
+  }
+
+  void from_json(const json &pars) override {
+    if (FieldExists(pars, "Weight")) {
+      weight_ = pars["Weight"];
+    } else {
+      weight_.setZero();
+    }
+    if (FieldExists(pars, "Bias")) {
+      bias_ = pars["Bias"];
+    } else {
+      bias_.setZero();
+    }
+  }
+
+  void InitRandomPars(int seed, double sigma) override {
+    VectorType par(npar_);
+
+    netket::RandomGaussian(par, seed, sigma);
+
+    SetParameters(par, 0);
+  }
+
+  int Npar() const override { return npar_; }
+
+  int Ninput() const override { return in_size_; }
+
+  int Noutput() const override { return out_size_; }
+
+  void GetParameters(VectorType &pars, int start_idx) override {
+    int k = start_idx;
+
+    if (usebias_) {
+      std::memcpy(pars.data() + k, bias_.data(), out_size_ * scalar_bytesize_);
+      k += out_size_;
+    }
+
+    std::memcpy(pars.data() + k, weight_.data(),
+                in_size_ * out_size_ * scalar_bytesize_);
+  }
+
+  void SetParameters(const VectorType &pars, int start_idx) override {
+    int k = start_idx;
+
+    if (usebias_) {
+      std::memcpy(bias_.data(), pars.data() + k, out_size_ * scalar_bytesize_);
+
+      k += out_size_;
+    }
+
+    std::memcpy(weight_.data(), pars.data() + k,
+                in_size_ * out_size_ * scalar_bytesize_);
+  }
+
+  void InitLookup(const Eigen::VectorXd &v, LookupType &lt) override {
+    if (lt.VectorSize() == 0) {
+      lt.AddVector(out_size_);
+    }
+    lt.V(0) = (weight_.transpose() * v + bias_);
+  }
+
+  void UpdateLookup(const Eigen::VectorXd &v, const std::vector<int> &tochange,
+                    const std::vector<double> &newconf,
+                    LookupType &lt) override {
+    if (tochange.size() != 0) {
+      for (std::size_t s = 0; s < tochange.size(); s++) {
+        const int sf = tochange[s];
+        lt.V(0) += weight_.row(sf) * (newconf[s] - v(sf));
+      }
+    }
+  }
+
+  void Forward(const VectorType &prev_layer_output,
+               VectorType &output) override {
+    // Linear term z = W' * in + b
+    z_ = bias_;
+    z_.noalias() += weight_.transpose() * prev_layer_output;
+
+    // Apply activation function
+    output.resize(out_size_);
+    activation_(z_, output);
+  }
+
+  // Using lookup
+  void Forward(const VectorType & /*prev_layer_output*/, const LookupType &lt,
+               VectorType &output) override {
+    z_ = lt.V(0);
+    // Apply activation function
+    output.resize(out_size_);
+    activation_(z_, output);
```

Comment:
At this point I would just do 
```c++
activation_(lt.V(0),output);
```
maybe resizing first if activation does not resize by default, or not resizing if output is already supposed to be of the good size

Reply:
actually depending on the activation function, sometimes it is not necessary to use z_